lib/Parse/ParseDecl.cpp

```diff
@@ -3703,14 +3728,18 @@ void Parser::ParseStructUnionBody(SourceLocation RecordLoc,
     }
 
     if (!Tok.is(tok::at)) {
-      auto CFieldCallback = [&](ParsingFieldDeclarator &FD) {
+      auto CFieldCallback = [&](ParsingFieldDeclarator &FD,
+                                CachedTokens *BoundsExprTokens) {
         // Install the declarator into the current TagDecl.
         Decl *Field =
             Actions.ActOnField(getCurScope(), TagDecl,
                                FD.D.getDeclSpec().getSourceRange().getBegin(),
                                FD.D, FD.BitfieldSize);
         FieldDecls.push_back(Field);
         FD.complete(Field);
+        if (BoundsExprTokens != nullptr) {
+          deferredBoundsExpressions.emplace_back(Field, BoundsExprTokens);
```

Comment:
>emplace_back [](start = 36, length = 12)

This is capturing pointers to the bounds expr tokens into the vector that outlasts the lifetime of the invocation to ParseStructDeclaration().  What's the contract around how long these tokens stay alive, who frees them, etc.?  This is precarious enough that I'd at least suggest that we carefully document how it's supposed to work.  If you have alternative ideas that allow a greater decoupling, I'd be all for that too.